{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 306,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>Acc. Length</th>\n",
       "      <th>sampleID</th>\n",
       "      <th>tool</th>\n",
       "      <th>gene</th>\n",
       "      <th>species</th>\n",
       "      <th>virus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>2</td>\n",
       "      <td>39</td>\n",
       "      <td>ERR3953710</td>\n",
       "      <td>VORF</td>\n",
       "      <td>gag protein, partial</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>6766</td>\n",
       "      <td>934</td>\n",
       "      <td>IVA</td>\n",
       "      <td>polymerase</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1782</td>\n",
       "      <td>934</td>\n",
       "      <td>IVA</td>\n",
       "      <td>nucleoprotein</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>4993</td>\n",
       "      <td>1880</td>\n",
       "      <td>IVA</td>\n",
       "      <td>polymerase</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>2238</td>\n",
       "      <td>1880</td>\n",
       "      <td>IVA</td>\n",
       "      <td>polymerase</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>1166</td>\n",
       "      <td>1880</td>\n",
       "      <td>IVA</td>\n",
       "      <td>glycoprotein precursor, partial</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4</td>\n",
       "      <td>1177</td>\n",
       "      <td>1880</td>\n",
       "      <td>IVA</td>\n",
       "      <td>nucleoprotein</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5</td>\n",
       "      <td>918</td>\n",
       "      <td>1880</td>\n",
       "      <td>IVA</td>\n",
       "      <td>polymerase</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>6</td>\n",
       "      <td>750</td>\n",
       "      <td>1880</td>\n",
       "      <td>IVA</td>\n",
       "      <td>glycoprotein precursor, partial</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7</td>\n",
       "      <td>714</td>\n",
       "      <td>1880</td>\n",
       "      <td>IVA</td>\n",
       "      <td>nucleoprotein</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>5605</td>\n",
       "      <td>540</td>\n",
       "      <td>IVA</td>\n",
       "      <td>polymerase</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2</td>\n",
       "      <td>1797</td>\n",
       "      <td>540</td>\n",
       "      <td>IVA</td>\n",
       "      <td>nucleoprotein</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>3</td>\n",
       "      <td>716</td>\n",
       "      <td>540</td>\n",
       "      <td>IVA</td>\n",
       "      <td>glycoprotein precursor, partial</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4</td>\n",
       "      <td>1208</td>\n",
       "      <td>540</td>\n",
       "      <td>IVA</td>\n",
       "      <td>polymerase</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "      <td>Lassa mammarenavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>4390</td>\n",
       "      <td>CV167</td>\n",
       "      <td>IVA</td>\n",
       "      <td>ORF1a polyprotein, partial</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>3</td>\n",
       "      <td>1278</td>\n",
       "      <td>CV167</td>\n",
       "      <td>IVA</td>\n",
       "      <td>ORF1a polyprotein, partial</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>5</td>\n",
       "      <td>1397</td>\n",
       "      <td>CV167</td>\n",
       "      <td>IVA</td>\n",
       "      <td>ORF1a polyprotein, partial</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>7</td>\n",
       "      <td>1173</td>\n",
       "      <td>CV167</td>\n",
       "      <td>IVA</td>\n",
       "      <td>ORF1a polyprotein, partial</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>10</td>\n",
       "      <td>900</td>\n",
       "      <td>CV167</td>\n",
       "      <td>IVA</td>\n",
       "      <td>ORF1a polyprotein, partial</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>11</td>\n",
       "      <td>890</td>\n",
       "      <td>CV167</td>\n",
       "      <td>IVA</td>\n",
       "      <td>ORF1a polyprotein, partial</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   contig  Acc. Length    sampleID  tool                             gene  \\\n",
       "70      2           39  ERR3953710  VORF             gag protein, partial   \n",
       "0       1         6766         934   IVA                       polymerase   \n",
       "1       2         1782         934   IVA                    nucleoprotein   \n",
       "2       1         4993        1880   IVA                       polymerase   \n",
       "3       2         2238        1880   IVA                       polymerase   \n",
       "4       3         1166        1880   IVA  glycoprotein precursor, partial   \n",
       "5       4         1177        1880   IVA                    nucleoprotein   \n",
       "6       5          918        1880   IVA                       polymerase   \n",
       "7       6          750        1880   IVA  glycoprotein precursor, partial   \n",
       "8       7          714        1880   IVA                    nucleoprotein   \n",
       "9       1         5605         540   IVA                       polymerase   \n",
       "10      2         1797         540   IVA                    nucleoprotein   \n",
       "11      3          716         540   IVA  glycoprotein precursor, partial   \n",
       "12      4         1208         540   IVA                       polymerase   \n",
       "13      1         4390       CV167   IVA       ORF1a polyprotein, partial   \n",
       "14      3         1278       CV167   IVA       ORF1a polyprotein, partial   \n",
       "15      5         1397       CV167   IVA       ORF1a polyprotein, partial   \n",
       "16      7         1173       CV167   IVA       ORF1a polyprotein, partial   \n",
       "17     10          900       CV167   IVA       ORF1a polyprotein, partial   \n",
       "18     11          890       CV167   IVA       ORF1a polyprotein, partial   \n",
       "\n",
       "                                            species  \\\n",
       "70                   Human immunodeficiency virus 1   \n",
       "0                              Lassa mammarenavirus   \n",
       "1                              Lassa mammarenavirus   \n",
       "2                              Lassa mammarenavirus   \n",
       "3                              Lassa mammarenavirus   \n",
       "4                              Lassa mammarenavirus   \n",
       "5                              Lassa mammarenavirus   \n",
       "6                              Lassa mammarenavirus   \n",
       "7                              Lassa mammarenavirus   \n",
       "8                              Lassa mammarenavirus   \n",
       "9                              Lassa mammarenavirus   \n",
       "10                             Lassa mammarenavirus   \n",
       "11                             Lassa mammarenavirus   \n",
       "12                             Lassa mammarenavirus   \n",
       "13  Severe acute respiratory syndrome coronavirus 2   \n",
       "14  Severe acute respiratory syndrome coronavirus 2   \n",
       "15  Severe acute respiratory syndrome coronavirus 2   \n",
       "16  Severe acute respiratory syndrome coronavirus 2   \n",
       "17  Severe acute respiratory syndrome coronavirus 2   \n",
       "18  Severe acute respiratory syndrome coronavirus 2   \n",
       "\n",
       "                                              virus  \n",
       "70                   Human immunodeficiency virus 1  \n",
       "0                              Lassa mammarenavirus  \n",
       "1                              Lassa mammarenavirus  \n",
       "2                              Lassa mammarenavirus  \n",
       "3                              Lassa mammarenavirus  \n",
       "4                              Lassa mammarenavirus  \n",
       "5                              Lassa mammarenavirus  \n",
       "6                              Lassa mammarenavirus  \n",
       "7                              Lassa mammarenavirus  \n",
       "8                              Lassa mammarenavirus  \n",
       "9                              Lassa mammarenavirus  \n",
       "10                             Lassa mammarenavirus  \n",
       "11                             Lassa mammarenavirus  \n",
       "12                             Lassa mammarenavirus  \n",
       "13  Severe acute respiratory syndrome coronavirus 2  \n",
       "14  Severe acute respiratory syndrome coronavirus 2  \n",
       "15  Severe acute respiratory syndrome coronavirus 2  \n",
       "16  Severe acute respiratory syndrome coronavirus 2  \n",
       "17  Severe acute respiratory syndrome coronavirus 2  \n",
       "18  Severe acute respiratory syndrome coronavirus 2  "
      ]
     },
     "execution_count": 306,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tool_list=['VORF','IVA']\n",
    "file_ext='_blast_results_wspecie_names.csv'\n",
    "df=[]\n",
    "\n",
    "for tool in tool_list:\n",
    "    tb = pd.read_csv(tool+file_ext,sep=\";\")\n",
    "    df.append(tb)\n",
    "\n",
    "df=pd.concat(df) \\\n",
    "    .rename(columns={\"V1\": \"contig\", \"V8\": \"Acc. Length\", \"sample\":\"sampleID\"})\n",
    "\n",
    "old_df=df\n",
    "\n",
    "df=df.drop(list(df.filter(regex = 'V\\d+')), axis = 1)\n",
    "\n",
    "df=pd.concat([df,df.specie.str.extract(r'(?P<gene>.*) \\[(?P<species>.*)\\].*')],axis=1) \\\n",
    "    .drop(['specie'],axis=1) \\\n",
    "    .applymap(lambda x: x.strip() if isinstance(x, str) else x) \\\n",
    "    .replace('C167', 'CV167')\n",
    "\n",
    "\n",
    "df['virus']=None\n",
    "df.loc[df.sampleID.str.contains(\"ERR\"), \"virus\"]=\"Human immunodeficiency virus 1\"\n",
    "df.loc[df.sampleID.str.isnumeric(), \"virus\"]=\"Lassa mammarenavirus\"\n",
    "df.loc[df.sampleID.str.contains(\"CV\"), \"virus\"]=\"Severe acute respiratory syndrome coronavirus 2\"\n",
    "\n",
    "df.contig=df.contig.str.lstrip(\"contig.0\").fillna(df.contig)\n",
    "\n",
    "df.columns = df.columns.str.lstrip()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  contig  Acc. Length    sampleID  tool                    gene  \\\n",
      "0      0          491  ERR3953716  VORF    gag protein, partial   \n",
      "1      4           32  ERR3953716  VORF    gag protein, partial   \n",
      "2      2          569        1880  VORF           nucleoprotein   \n",
      "3      5          490        1880  VORF  glycoprotein precursor   \n",
      "4     14         2032         540  VORF               L protein   \n",
      "\n",
      "                          species Length  \n",
      "0  Human immunodeficiency virus 1    491  \n",
      "1  Human immunodeficiency virus 1     37  \n",
      "2            Lassa mammarenavirus    569  \n",
      "3            Lassa mammarenavirus    490  \n",
      "4            Lassa mammarenavirus   2048  \n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "df[\"Length\"] = None\n",
    "sample_IDs=df.sampleID.unique()\n",
    "\n",
    "contigs=pd.DataFrame({'sampleID': [],'tool': [], 'contig': [],'length':[]})\n",
    "\n",
    "for tool in tool_list:\n",
    "    for s in sample_IDs:\n",
    "        sid= (s + \"_VORF\") if tool==\"VORF\" else (s + \"_contigs\")\n",
    "        filename='Data/%s_assemblies/%s.fasta' % (tool, sid)\n",
    "        fasta_sequences = SeqIO.parse(open(filename),'fasta')\n",
    "        for fasta in fasta_sequences:\n",
    "            \n",
    "            name, sequence = fasta.id, str(fasta.seq)\n",
    "            length=len(sequence)\n",
    "            contigs.loc[len(contigs.index)] = [s,tool,name,length]\n",
    "            \n",
    "            name=int(name) if tool==\"VORF\" else name.lstrip('contig.0')\n",
    "            df.loc[(df.contig==name) & (df.sampleID==s), [\"Length\"]]=length\n",
    "\n",
    "\n",
    "contigs['virus']=None\n",
    "contigs.loc[contigs.sampleID.str.contains(\"ERR\"), \"virus\"]=\"HIV\"\n",
    "contigs.loc[contigs.sampleID.str.isnumeric(), \"virus\"]=\"Lassa\"\n",
    "contigs.loc[contigs.sampleID.str.contains(\"CV\"), \"virus\"]=\"SARS-CoV-2\"\n",
    "\n",
    "print(df.head())\n",
    "\n",
    "\n",
    "\n",
    "df.to_csv(\"blast_dataset.csv\",sep=\";\")\n",
    "contigs.to_csv(\"contigs.csv\",sep=\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th colspan=\"2\" halign=\"left\">length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>sum</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tool</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IVA</th>\n",
       "      <td>97</td>\n",
       "      <td>1891.896907</td>\n",
       "      <td>183514.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VORF</th>\n",
       "      <td>223</td>\n",
       "      <td>77.394619</td>\n",
       "      <td>17259.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     contig       length          \n",
       "      count         mean       sum\n",
       "tool                              \n",
       "IVA      97  1891.896907  183514.0\n",
       "VORF    223    77.394619   17259.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th colspan=\"2\" halign=\"left\">length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>sum</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tool</th>\n",
       "      <th>virus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">IVA</th>\n",
       "      <th>HIV</th>\n",
       "      <td>11</td>\n",
       "      <td>1152.272727</td>\n",
       "      <td>12675.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lassa</th>\n",
       "      <td>13</td>\n",
       "      <td>2640.923077</td>\n",
       "      <td>34332.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SARS-CoV-2</th>\n",
       "      <td>73</td>\n",
       "      <td>1869.958904</td>\n",
       "      <td>136507.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">VORF</th>\n",
       "      <th>HIV</th>\n",
       "      <td>25</td>\n",
       "      <td>84.400000</td>\n",
       "      <td>2110.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lassa</th>\n",
       "      <td>80</td>\n",
       "      <td>110.325000</td>\n",
       "      <td>8826.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SARS-CoV-2</th>\n",
       "      <td>118</td>\n",
       "      <td>53.584746</td>\n",
       "      <td>6323.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                contig       length          \n",
       "                 count         mean       sum\n",
       "tool virus                                   \n",
       "IVA  HIV            11  1152.272727   12675.0\n",
       "     Lassa          13  2640.923077   34332.0\n",
       "     SARS-CoV-2     73  1869.958904  136507.0\n",
       "VORF HIV            25    84.400000    2110.0\n",
       "     Lassa          80   110.325000    8826.0\n",
       "     SARS-CoV-2    118    53.584746    6323.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display\n",
    "# average # of contigs identified by each tool\n",
    "\n",
    "display(contigs.groupby(['tool']).agg({'contig':['count'], 'length':['mean', 'sum']}))\n",
    "\n",
    "display(contigs.groupby(['tool', 'virus']).agg({'contig':['count'], 'length':['mean', 'sum']}))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# average # of contigs identified by each tool in each sample\n",
    "\n",
    "# average length of contigs identified by each tool\n",
    "\n",
    "# average length of contigs identified by each tool in each sample\n",
    "\n",
    "# total length of all segments identified by each tool\n",
    "\n",
    "# correlation between # of reads and segments found\n",
    "\n",
    "#  number of base pairs aligned to parnet genome\n",
    "\n",
    "# base pairs aligned to bacterial genomes\n",
    "\n",
    "# base pairs aligned to nothing\n",
    "\n",
    "# proteins identified by each tool in each sample\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>virus</th>\n",
       "      <th>species</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Human immunodeficiency virus 1</th>\n",
       "      <th>Human immunodeficiency virus 1</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lassa mammarenavirus</th>\n",
       "      <th>Lassa mammarenavirus</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                               species\n",
       "virus                          species                                \n",
       "Human immunodeficiency virus 1 Human immunodeficiency virus 1        8\n",
       "Lassa mammarenavirus           Lassa mammarenavirus                  7"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>virus</th>\n",
       "      <th>species</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Human immunodeficiency virus 1</th>\n",
       "      <th>Human immunodeficiency virus 1</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lassa mammarenavirus</th>\n",
       "      <th>Lassa mammarenavirus</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Severe acute respiratory syndrome coronavirus 2</th>\n",
       "      <th>Severe acute respiratory syndrome coronavirus 2</th>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                 species\n",
       "virus                                           species                                                 \n",
       "Human immunodeficiency virus 1                  Human immunodeficiency virus 1                        11\n",
       "Lassa mammarenavirus                            Lassa mammarenavirus                                  13\n",
       "Severe acute respiratory syndrome coronavirus 2 Severe acute respiratory syndrome coronavirus 2       36"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tool</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tool</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IVA</th>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VORF</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      tool\n",
       "tool      \n",
       "IVA     60\n",
       "VORF    15"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# number of contigs in a virus aligned to the right species\n",
    "\n",
    "display(df[(df.virus==df.species) & (df.tool==\"VORF\")].groupby(['virus', 'species']).agg({'species':'count'}))\n",
    "\n",
    "display(df[(df.virus==df.species) & (df.tool==\"IVA\")].groupby(['virus', 'species']).agg({'species':'count'}))\n",
    "\n",
    "display(df[df.virus==df.species].groupby(['tool']).agg({'tool':'count'}))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>virus</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>virus</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Lassa mammarenavirus</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Severe acute respiratory syndrome coronavirus 2</th>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 virus\n",
       "virus                                                 \n",
       "Lassa mammarenavirus                                 1\n",
       "Severe acute respiratory syndrome coronavirus 2     55"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# contigs aligned to bacterial genomes\n",
    "\n",
    "display(df[(df.virus!=df.species) & (df.tool==\"VORF\")].groupby(['virus']).agg({'virus':'count'}))"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
  },
  "kernelspec": {
   "display_name": "Python 3.9.1 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
